File: microservices/nllb-local/scripts/download-model.js
const fs = require('fs')
const path = require('path')
const https = require('https')
const { pipeline } = require('@huggingface/transformers')

class ModelDownloader {
  constructor() {
    this.modelName = 'facebook/nllb-200-distilled-600M'
    this.modelDir = path.join(__dirname, '../models/nllb-600m')
    this.downloadDir = path.join(__dirname, '../downloads')
  }

  /**
   * 确保目录存在
   */
  ensureDirectory(dir) {
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true })
      console.log(`Created directory: ${dir}`)
    }
  }

  /**
   * 下载模型文件
   */
  async downloadModel() {
    console.log('Starting NLLB 600M model download...')
    console.log(`Model: ${this.modelName}`)
    console.log(`Target directory: ${this.modelDir}`)

    try {
      this.ensureDirectory(this.modelDir)
      this.ensureDirectory(this.downloadDir)

      // 使用transformers.js下载并缓存模型
      console.log('Downloading model from Hugging Face Hub...')
      
      const model = await pipeline('translation', this.modelName, {
        cache_dir: this.downloadDir
      })

      console.log('Model downloaded successfully!')
      console.log(`Model cached in: ${this.downloadDir}`)
      
      // 验证下载
      await this.verifyDownload()
      
      return true

    } catch (error) {
      console.error('Download failed:', error)
      throw error
    }
  }

  /**
   * 验证下载的模型
   */
  async verifyDownload() {
    console.log('Verifying downloaded model...')
    
    try {
      // 简单的验证：尝试加载模型并进行一次翻译
      const testModel = await pipeline('translation', this.modelName, {
        cache_dir: this.downloadDir
      })

      const testResult = await testModel('Hello world', {
        src_lang: 'eng_Latn',
        tgt_lang: 'fra_Latn'
      })

      console.log('Verification successful!')
      console.log('Test translation:', testResult)
      return true

    } catch (error) {
      console.error('Verification failed:', error)
      throw error
    }
  }

  /**
   * 获取模型信息
   */
  async getModelInfo() {
    const modelSize = await this.getDirectorySize(this.downloadDir)
    
    return {
      modelName: this.modelName,
      downloadDir: this.downloadDir,
      modelDir: this.modelDir,
      sizeGB: (modelSize / (1024 * 1024 * 1024)).toFixed(2),
      downloaded: fs.existsSync(this.downloadDir)
    }
  }

  /**
   * 计算目录大小
   */
  async getDirectorySize(dirPath) {
    if (!fs.existsSync(dirPath)) return 0

    let totalSize = 0
    const files = fs.readdirSync(dirPath)

    for (const file of files) {
      const filePath = path.join(dirPath, file)
      const stats = fs.statSync(filePath)

      if (stats.isDirectory()) {
        totalSize += await this.getDirectorySize(filePath)
      } else {
        totalSize += stats.size
      }
    }

    return totalSize
  }

  /**
   * 清理下载文件
   */
  cleanup() {
    if (fs.existsSync(this.downloadDir)) {
      fs.rmSync(this.downloadDir, { recursive: true, force: true })
      console.log('Download cache cleaned up')
    }
  }
}

// 命令行使用
async function main() {
  const downloader = new ModelDownloader()
  
  const command = process.argv[2] || 'download'

  try {
    switch (command) {
      case 'download':
        await downloader.downloadModel()
        break
      
      case 'info':
        const info = await downloader.getModelInfo()
        console.log('Model Information:')
        console.log(`  Name: ${info.modelName}`)
        console.log(`  Download Directory: ${info.downloadDir}`)
        console.log(`  Model Directory: ${info.modelDir}`)
        console.log(`  Size: ${info.sizeGB} GB`)
        console.log(`  Downloaded: ${info.downloaded}`)
        break
      
      case 'cleanup':
        downloader.cleanup()
        break
      
      case 'verify':
        await downloader.verifyDownload()
        break
      
      default:
        console.log('Usage: node download-model.js [download|info|cleanup|verify]')
        break
    }
  } catch (error) {
    console.error('Command failed:', error)
    process.exit(1)
  }
}

if (require.main === module) {
  main()
}

module.exports = ModelDownloader 

File: microservices/nllb-local/scripts/download_model.py
#!/usr/bin/env python3
"""
NLLB Model Download and Management Script
Downloads and manages the NLLB 600M model for local inference
"""

import os
import sys
import argparse
from pathlib import Path

def check_dependencies():
    """检查Python依赖"""
    required_packages = ['transformers', 'torch', 'sentencepiece']
    missing_packages = []
    
    for package in required_packages:
        try:
            __import__(package)
        except ImportError:
            missing_packages.append(package)
    
    if missing_packages:
        print(f"❌ Missing required packages: {', '.join(missing_packages)}")
        print("Installing required packages...")
        
        import subprocess
        for package in missing_packages:
            subprocess.check_call([sys.executable, "-m", "pip", "install", package])
        
        print("✅ All packages installed successfully!")

def download_model():
    """下载NLLB模型"""
    print("🚀 Starting NLLB 600M model download...")
    
    # 检查依赖
    check_dependencies()
    
    try:
        from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
        
        model_name = "facebook/nllb-200-distilled-600M"
        model_dir = Path(__file__).parent.parent / "models" / "nllb-600m"
        
        print(f"📁 Model directory: {model_dir}")
        model_dir.mkdir(parents=True, exist_ok=True)
        
        # 下载tokenizer
        print("📥 Downloading tokenizer...")
        tokenizer = AutoTokenizer.from_pretrained(model_name)
        tokenizer.save_pretrained(model_dir)
        
        # 下载模型
        print("📥 Downloading model (this may take a while)...")
        model = AutoModelForSeq2SeqLM.from_pretrained(model_name)
        model.save_pretrained(model_dir)
        
        print("✅ Model downloaded successfully!")
        print(f"📁 Model saved to: {model_dir}")
        
        # 验证下载
        verify_model()
        
    except Exception as e:
        print(f"❌ Download failed: {e}")
        return False
    
    return True

def verify_model():
    """验证模型"""
    print("🔍 Verifying model...")
    
    try:
        from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
        
        model_dir = Path(__file__).parent.parent / "models" / "nllb-600m"
        
        # 加载模型和tokenizer
        tokenizer = AutoTokenizer.from_pretrained(model_dir)
        model = AutoModelForSeq2SeqLM.from_pretrained(model_dir)
        
        # 测试翻译
        test_text = "Hello world"
        inputs = tokenizer(test_text, return_tensors="pt")
        outputs = model.generate(**inputs, forced_bos_token_id=tokenizer.convert_tokens_to_ids("hat_Latn"))
        result = tokenizer.decode(outputs[0], skip_special_tokens=True)
        
        print(f"✅ Model verification successful!")
        print(f"🧪 Test translation: '{test_text}' -> '{result}'")
        
    except Exception as e:
        print(f"❌ Model verification failed: {e}")
        return False
    
    return True

def get_model_info():
    """获取模型信息"""
    model_dir = Path(__file__).parent.parent / "models" / "nllb-600m"
    
    if not model_dir.exists():
        print("❌ Model not found. Run download first.")
        return
    
    # 计算模型大小
    total_size = sum(f.stat().st_size for f in model_dir.rglob('*') if f.is_file())
    size_gb = total_size / (1024 * 1024 * 1024)
    
    print("📊 Model Information:")
    print(f"  📁 Location: {model_dir}")
    print(f"  📦 Size: {size_gb:.2f} GB")
    print(f"  ✅ Status: {'Available' if model_dir.exists() else 'Not downloaded'}")

def cleanup_model():
    """清理模型文件"""
    model_dir = Path(__file__).parent.parent / "models" / "nllb-600m"
    downloads_dir = Path(__file__).parent.parent / "downloads"
    
    import shutil
    
    if model_dir.exists():
        shutil.rmtree(model_dir)
        print(f"🗑️  Removed model directory: {model_dir}")
    
    if downloads_dir.exists():
        shutil.rmtree(downloads_dir)
        print(f"🗑️  Removed downloads directory: {downloads_dir}")
    
    print("✅ Cleanup completed!")

def main():
    parser = argparse.ArgumentParser(description="NLLB Model Management")
    parser.add_argument("command", choices=["download", "verify", "info", "cleanup"], 
                       default="download", nargs="?",
                       help="Command to execute")
    
    args = parser.parse_args()
    
    if args.command == "download":
        download_model()
    elif args.command == "verify":
        verify_model()
    elif args.command == "info":
        get_model_info()
    elif args.command == "cleanup":
        cleanup_model()

if __name__ == "__main__":
    main() 